// @databricks/appkit/dist/plugins/server/index.js
import { createLogger } from "../../logging/logger.js";
import { ServerError } from "../../errors/server.js";
import { init_errors } from "../../errors/index.js";
import { instrumentations } from "../../telemetry/instrumentations.js";
import "../../telemetry/index.js";
import { Plugin } from "../../plugin/plugin.js";
import { toPlugin } from "../../plugin/to-plugin.js";
import "../../plugin/index.js";
import { serverManifest } from "./manifest.js";
import { RemoteTunnelController } from "./remote-tunnel/remote-tunnel-controller.js";
import { getRoutes } from "./utils.js";
import { StaticServer } from "./static-server.js";
import { ViteDevServer } from "./vite-dev-server.js";
import path from "node:path";
import fs from "node:fs";
import dotenv from "dotenv";
import express from "express";

//#region src/plugins/server/index.ts
init_errors();
dotenv.config({ path: path.resolve(process.cwd(), "./.env") });
const logger = createLogger("server");
/**
* Server plugin for the AppKit.
*
* This plugin is responsible for starting the server and serving the static files.
* It also handles the remote tunneling for development purposes.
*
* @example
* ```ts
* createApp({
*   plugins: [server(), telemetryExamples(), analytics({})],
* });
* ```
*
*/
var ServerPlugin = class ServerPlugin extends Plugin {
	static DEFAULT_CONFIG = {
		autoStart: true,
		host: process.env.FLASK_RUN_HOST || "0.0.0.0",
		port: Number(process.env.DATABRICKS_APP_PORT) || 8e3
	};
	/** Plugin manifest declaring metadata and resource requirements */
	static manifest = serverManifest;
	name = "server";
	serverApplication;
	server;
	viteDevServer;
	remoteTunnelController;
	serverExtensions = [];
	static phase = "deferred";
	constructor(config) {
		super(config);
		this.config = config;
		this.serverApplication = express();
		this.server = null;
		this.serverExtensions = [];
		this.telemetry.registerInstrumentations([instrumentations.http, instrumentations.express]);
	}
	/** Setup the server plugin. */
	async setup() {
		if (this.shouldAutoStart()) await this.start();
	}
	/** Get the server configuration. */
	getConfig() {
		const { plugins: _plugins, ...config } = this.config;
		return config;
	}
	/** Check if the server should auto start. */
	shouldAutoStart() {
		return this.config.autoStart;
	}
	/**
	* Start the server.
	*
	* This method starts the server and sets up the frontend.
	* It also sets up the remote tunneling if enabled.
	*
	* @returns The express application.
	*/
	async start() {
		this.serverApplication.use(express.json());
		const endpoints = await this.extendRoutes();
		for (const extension of this.serverExtensions) extension(this.serverApplication);
		this.remoteTunnelController = new RemoteTunnelController(this.devFileReader);
		this.serverApplication.use(this.remoteTunnelController.middleware);
		await this.setupFrontend(endpoints);
		const server = this.serverApplication.listen(this.config.port ?? ServerPlugin.DEFAULT_CONFIG.port, this.config.host ?? ServerPlugin.DEFAULT_CONFIG.host, () => this.logStartupInfo());
		this.server = server;
		this.remoteTunnelController.setServer(server);
		process.on("SIGTERM", () => this._gracefulShutdown());
		process.on("SIGINT", () => this._gracefulShutdown());
		if (process.env.NODE_ENV === "development") {
			const allRoutes = getRoutes(this.serverApplication._router.stack);
			console.dir(allRoutes, { depth: null });
		}
		return this.serverApplication;
	}
	/**
	* Get the low level node.js http server instance.
	*
	* Only use this method if you need to access the server instance for advanced usage like a custom websocket server, etc.
	*
	* @throws {Error} If the server is not started or autoStart is true.
	* @returns {HTTPServer} The server instance.
	*/
	getServer() {
		if (this.shouldAutoStart()) throw ServerError.autoStartConflict("get server");
		if (!this.server) throw ServerError.notStarted();
		return this.server;
	}
	/**
	* Extend the server with custom routes or middleware.
	*
	* @param fn - A function that receives the express application.
	* @returns The server plugin instance for chaining.
	* @throws {Error} If autoStart is true.
	*/
	extend(fn) {
		if (this.shouldAutoStart()) throw ServerError.autoStartConflict("extend server");
		this.serverExtensions.push(fn);
		return this;
	}
	/**
	* Setup the routes with the plugins.
	*
	* This method goes through all the plugins and injects the routes into the server application.
	* Returns a map of plugin names to their registered named endpoints.
	*/
	async extendRoutes() {
		const endpoints = {};
		if (!this.config.plugins) return endpoints;
		this.serverApplication.get("/health", (_, res) => {
			res.status(200).json({ status: "ok" });
		});
		this.registerEndpoint("health", "/health");
		for (const plugin of Object.values(this.config.plugins)) {
			if (EXCLUDED_PLUGINS.includes(plugin.name)) continue;
			if (plugin?.injectRoutes && typeof plugin.injectRoutes === "function") {
				const router = express.Router();
				plugin.injectRoutes(router);
				const basePath = `/api/${plugin.name}`;
				this.serverApplication.use(basePath, router);
				endpoints[plugin.name] = plugin.getEndpoints();
			}
		}
		return endpoints;
	}
	/**
	* Setup frontend serving based on environment:
	* - If staticPath is explicitly provided: use static server
	* - Dev mode (no staticPath): Vite for HMR
	* - Production (no staticPath): Static files auto-detected
	*/
	async setupFrontend(endpoints) {
		const isDev = process.env.NODE_ENV === "development";
		if (this.config.staticPath !== void 0) {
			new StaticServer(this.serverApplication, this.config.staticPath, endpoints).setup();
			return;
		}
		if (isDev) {
			this.viteDevServer = new ViteDevServer(this.serverApplication, endpoints);
			await this.viteDevServer.setup();
			return;
		}
		const staticPath = ServerPlugin.findStaticPath();
		if (staticPath) new StaticServer(this.serverApplication, staticPath, endpoints).setup();
	}
	static findStaticPath() {
		const staticPaths = [
			"dist",
			"client/dist",
			"build",
			"public",
			"out"
		];
		const cwd = process.cwd();
		for (const p of staticPaths) {
			const fullPath = path.resolve(cwd, p);
			if (fs.existsSync(path.resolve(fullPath, "index.html"))) {
				logger.debug("Static files: serving from %s", fullPath);
				return fullPath;
			}
		}
	}
	logStartupInfo() {
		const isDev = process.env.NODE_ENV === "development";
		const hasExplicitStaticPath = this.config.staticPath !== void 0;
		const port = this.config.port ?? ServerPlugin.DEFAULT_CONFIG.port;
		const host = this.config.host ?? ServerPlugin.DEFAULT_CONFIG.host;
		logger.info("Server running on http://%s:%d", host, port);
		if (hasExplicitStaticPath) logger.info("Mode: static (%s)", this.config.staticPath);
		else if (isDev) logger.info("Mode: development (Vite HMR)");
		else logger.info("Mode: production (static)");
		const remoteServerController = this.remoteTunnelController;
		if (!remoteServerController) logger.debug("Remote tunnel: disabled (controller not initialized)");
		else logger.debug("Remote tunnel: %s; %s", remoteServerController.isAllowedByEnv() ? "allowed" : "blocked", remoteServerController.isActive() ? "active" : "inactive");
	}
	async _gracefulShutdown() {
		logger.info("Starting graceful shutdown...");
		if (this.viteDevServer) await this.viteDevServer.close();
		if (this.remoteTunnelController) this.remoteTunnelController.cleanup();
		if (this.config.plugins) {
			for (const plugin of Object.values(this.config.plugins)) if (plugin.abortActiveOperations) try {
				plugin.abortActiveOperations();
			} catch (err) {
				logger.error("Error aborting operations for plugin %s: %O", plugin.name, err);
			}
		}
		if (this.server) {
			this.server.close(() => {
				logger.debug("Server closed gracefully");
				process.exit(0);
			});
			setTimeout(() => {
				logger.debug("Force shutdown after timeout");
				process.exit(1);
			}, 15e3);
		} else process.exit(0);
	}
	/**
	* Returns the public exports for the server plugin.
	* Exposes server management methods.
	*/
	exports() {
		const self = this;
		return {
			start: this.start,
			extend(fn) {
				self.extend(fn);
				return this;
			},
			getServer: this.getServer,
			getConfig: this.getConfig
		};
	}
};
const EXCLUDED_PLUGINS = [ServerPlugin.name];
/**
* @internal
*/
const server = toPlugin(ServerPlugin, "server");

//#endregion
export { ServerPlugin, server };
