// @databricks/appkit/dist/plugins/server/utils.js
import crypto from "node:crypto";
import path from "node:path";
import fs from "node:fs";

//#region src/plugins/server/utils.ts
function parseCookies(req) {
	const cookieHeader = req.headers.cookie;
	if (!cookieHeader) return {};
	if (cookieHeader.indexOf(";") === -1) {
		const eqIndex = cookieHeader.indexOf("=");
		if (eqIndex === -1) return {};
		return { [cookieHeader.slice(0, eqIndex).trim()]: cookieHeader.slice(eqIndex + 1) };
	}
	const cookies = {};
	const parts = cookieHeader.split(";");
	for (let i = 0; i < parts.length; i++) {
		const eqIndex = parts[i].indexOf("=");
		if (eqIndex !== -1) {
			const key = parts[i].slice(0, eqIndex).trim();
			cookies[key] = parts[i].slice(eqIndex + 1);
		}
	}
	return cookies;
}
function generateTunnelIdFromEmail(email) {
	if (!email) return void 0;
	return crypto.createHash("sha256").update(email).digest("base64url").slice(0, 8);
}
function getRoutes(stack, basePath = "") {
	const routes = [];
	stack.forEach((layer) => {
		if (layer.route) {
			const path = basePath + layer.route.path;
			const methods = Object.keys(layer.route.methods).map((m) => m.toUpperCase());
			routes.push({
				path,
				methods
			});
		} else if (layer.name === "router" && layer.handle.stack) {
			const nestedBase = basePath + layer.regexp.source.replace("^\\", "").replace("\\/?(?=\\/|$)", "").replace(/\\\//g, "/").replace(/\$$/, "") || "";
			routes.push(...getRoutes(layer.handle.stack, nestedBase));
		}
	});
	return routes;
}
function getQueries(configFolder) {
	const queriesFolder = path.join(configFolder, "queries");
	if (!fs.existsSync(queriesFolder)) return {};
	return Object.fromEntries(fs.readdirSync(queriesFolder).filter((f) => path.extname(f) === ".sql").map((f) => [path.basename(f, ".sql"), path.basename(f, ".sql")]));
}
function getRuntimeConfig(endpoints = {}) {
	const configFolder = path.join(process.cwd(), "config");
	return {
		appName: process.env.DATABRICKS_APP_NAME || "",
		queries: getQueries(configFolder),
		endpoints
	};
}
function getConfigScript(endpoints = {}) {
	const config = getRuntimeConfig(endpoints);
	return `
    <script>
      window.__CONFIG__ = ${JSON.stringify(config)};
    <\/script>
  `;
}

//#endregion
export { generateTunnelIdFromEmail, getConfigScript, getRoutes, parseCookies };
