// @databricks/appkit-ui/dist/react/charts/options.js
import { createTimeSeriesData, formatLabel, truncateLabel } from "./utils.js";

//#region src/react/charts/options.ts
function buildBaseOption(ctx) {
	return {
		title: ctx.title ? {
			text: ctx.title,
			left: "center"
		} : void 0,
		color: ctx.colors
	};
}
function buildRadarOption(ctx, showArea = true) {
	const maxValue = Math.max(...ctx.yFields.flatMap((f) => ctx.yDataMap[f].map((v) => Number(v) || 0)));
	return {
		...buildBaseOption(ctx),
		tooltip: { trigger: "item" },
		legend: ctx.showLegend && ctx.yFields.length > 1 ? { top: "bottom" } : void 0,
		radar: {
			indicator: ctx.xData.map((name) => ({
				name: String(name),
				max: maxValue * 1.2
			})),
			shape: "polygon"
		},
		series: [{
			type: "radar",
			data: ctx.yFields.map((key, idx) => ({
				name: formatLabel(key),
				value: ctx.yDataMap[key],
				itemStyle: { color: ctx.colors[idx % ctx.colors.length] },
				areaStyle: showArea ? { opacity: .3 } : void 0
			}))
		}]
	};
}
function buildPieOption(ctx, chartType, innerRadius, showLabels, labelPosition) {
	const pieData = ctx.xData.map((name, i) => ({
		name: String(name),
		value: ctx.yDataMap[ctx.yFields[0]]?.[i] ?? 0
	}));
	const isDonut = chartType === "donut" || innerRadius > 0;
	return {
		...buildBaseOption(ctx),
		tooltip: {
			trigger: "item",
			formatter: "{b}: {c} ({d}%)"
		},
		legend: ctx.showLegend ? {
			orient: "vertical",
			left: "left",
			top: "middle"
		} : void 0,
		series: [{
			type: "pie",
			radius: isDonut ? [`${innerRadius || 40}%`, "70%"] : "70%",
			center: ["60%", "50%"],
			data: pieData,
			label: {
				show: showLabels,
				position: labelPosition,
				formatter: "{b}: {d}%"
			},
			emphasis: { itemStyle: {
				shadowBlur: 10,
				shadowOffsetX: 0,
				shadowColor: "rgba(0, 0, 0, 0.5)"
			} }
		}]
	};
}
function buildHorizontalBarOption(ctx, stacked) {
	const hasMultipleSeries = ctx.yFields.length > 1;
	return {
		...buildBaseOption(ctx),
		tooltip: {
			trigger: "axis",
			axisPointer: { type: "shadow" }
		},
		legend: ctx.showLegend && hasMultipleSeries ? { top: "bottom" } : void 0,
		grid: {
			left: "20%",
			right: "10%",
			top: ctx.title ? "15%" : "5%",
			bottom: ctx.showLegend && hasMultipleSeries ? "15%" : "5%"
		},
		xAxis: { type: "value" },
		yAxis: {
			type: "category",
			data: ctx.xData,
			axisLabel: {
				width: 100,
				overflow: "truncate",
				formatter: (value) => truncateLabel(String(value))
			}
		},
		series: ctx.yFields.map((key, idx) => ({
			name: formatLabel(key),
			type: "bar",
			data: ctx.yDataMap[key],
			stack: stacked ? "total" : void 0,
			itemStyle: { borderRadius: [
				0,
				4,
				4,
				0
			] },
			color: ctx.colors[idx % ctx.colors.length]
		}))
	};
}
function buildHeatmapOption(ctx) {
	return {
		...buildBaseOption(ctx),
		tooltip: {
			trigger: "item",
			formatter: (params) => {
				const [xIdx, yIdx, value] = params.data;
				return `${ctx.xData[xIdx] ?? xIdx}, ${ctx.yAxisData[yIdx] ?? yIdx}: ${value}`;
			}
		},
		grid: {
			left: "15%",
			right: "15%",
			top: ctx.title ? "15%" : "10%",
			bottom: "15%"
		},
		xAxis: {
			type: "category",
			data: ctx.xData,
			splitArea: { show: true },
			axisLabel: {
				rotate: ctx.xData.length > 10 ? 45 : 0,
				formatter: (v) => truncateLabel(String(v), 10)
			}
		},
		yAxis: {
			type: "category",
			data: ctx.yAxisData,
			splitArea: { show: true },
			axisLabel: { formatter: (v) => truncateLabel(String(v), 12) }
		},
		visualMap: {
			min: ctx.min,
			max: ctx.max,
			calculable: true,
			orient: "vertical",
			right: "2%",
			top: "center",
			inRange: { color: ctx.colors.length >= 2 ? ctx.colors : ["#f0f0f0", ctx.colors[0]] }
		},
		series: [{
			type: "heatmap",
			data: ctx.heatmapData,
			label: {
				show: ctx.showLabels,
				formatter: (params) => String(params.data[2])
			},
			emphasis: { itemStyle: {
				shadowBlur: 10,
				shadowColor: "rgba(0, 0, 0, 0.5)"
			} }
		}]
	};
}
function buildCartesianOption(ctx) {
	const { chartType, isTimeSeries, stacked, smooth, showSymbol, symbolSize } = ctx;
	const hasMultipleSeries = ctx.yFields.length > 1;
	const seriesType = chartType === "area" ? "line" : chartType;
	return {
		...buildBaseOption(ctx),
		tooltip: { trigger: "axis" },
		legend: ctx.showLegend && hasMultipleSeries ? { top: "bottom" } : void 0,
		grid: {
			left: "10%",
			right: "10%",
			top: ctx.title ? "15%" : "10%",
			bottom: ctx.showLegend && hasMultipleSeries ? "20%" : "15%"
		},
		xAxis: {
			type: isTimeSeries ? "time" : "category",
			data: isTimeSeries ? void 0 : ctx.xData,
			axisLabel: isTimeSeries ? void 0 : {
				rotate: ctx.xData.length > 10 ? 45 : 0,
				formatter: (v) => truncateLabel(String(v), 10)
			}
		},
		yAxis: { type: "value" },
		series: ctx.yFields.map((key, idx) => ({
			name: formatLabel(key),
			type: seriesType,
			data: isTimeSeries ? createTimeSeriesData(ctx.xData, ctx.yDataMap[key]) : ctx.yDataMap[key],
			smooth: chartType === "line" || chartType === "area" ? smooth : void 0,
			showSymbol: chartType === "line" || chartType === "area" ? showSymbol : void 0,
			symbol: chartType === "scatter" ? "circle" : void 0,
			symbolSize: chartType === "scatter" ? symbolSize : void 0,
			areaStyle: chartType === "area" ? { opacity: .3 } : void 0,
			stack: stacked && chartType === "area" ? "total" : void 0,
			itemStyle: chartType === "bar" ? { borderRadius: [
				4,
				4,
				0,
				0
			] } : void 0,
			color: ctx.colors[idx % ctx.colors.length]
		}))
	};
}

//#endregion
export { buildCartesianOption, buildHeatmapOption, buildHorizontalBarOption, buildPieOption, buildRadarOption };
