// @databricks/appkit/dist/plugins/analytics/analytics.js
import { createLogger } from "../../logging/logger.js";
import { getCurrentUserId, getWarehouseId, getWorkspaceClient } from "../../context/execution-context.js";
import { init_context } from "../../context/index.js";
import { Plugin } from "../../plugin/plugin.js";
import { toPlugin } from "../../plugin/to-plugin.js";
import "../../plugin/index.js";
import { SQLWarehouseConnector } from "../../connectors/sql-warehouse/client.js";
import "../../connectors/index.js";
import { queryDefaults } from "./defaults.js";
import { analyticsManifest } from "./manifest.js";
import { QueryProcessor } from "./query.js";

//#region src/plugins/analytics/analytics.ts
init_context();
const logger = createLogger("analytics");
var AnalyticsPlugin = class extends Plugin {
	name = "analytics";
	/** Plugin manifest declaring metadata and resource requirements */
	static manifest = analyticsManifest;
	static description = "Analytics plugin for data analysis";
	SQLClient;
	queryProcessor;
	constructor(config) {
		super(config);
		this.config = config;
		this.queryProcessor = new QueryProcessor();
		this.SQLClient = new SQLWarehouseConnector({
			timeout: config.timeout,
			telemetry: config.telemetry
		});
	}
	injectRoutes(router) {
		this.route(router, {
			name: "arrow",
			method: "get",
			path: "/arrow-result/:jobId",
			handler: async (req, res) => {
				await this._handleArrowRoute(req, res);
			}
		});
		this.route(router, {
			name: "query",
			method: "post",
			path: "/query/:query_key",
			handler: async (req, res) => {
				await this._handleQueryRoute(req, res);
			}
		});
	}
	/**
	* Handle Arrow data download requests.
	* When called via asUser(req), uses the user's Databricks credentials.
	*/
	async _handleArrowRoute(req, res) {
		try {
			const { jobId } = req.params;
			const workspaceClient = getWorkspaceClient();
			logger.debug("Processing Arrow job request for jobId=%s", jobId);
			logger.event(req)?.setComponent("analytics", "getArrowData").setContext("analytics", {
				job_id: jobId,
				plugin: this.name
			});
			const result = await this.getArrowData(workspaceClient, jobId);
			res.setHeader("Content-Type", "application/octet-stream");
			res.setHeader("Content-Length", result.data.length.toString());
			res.setHeader("Cache-Control", "public, max-age=3600");
			logger.debug("Sending Arrow buffer: %d bytes for job %s", result.data.length, jobId);
			res.send(Buffer.from(result.data));
		} catch (error) {
			logger.error("Arrow job error: %O", error);
			res.status(404).json({
				error: error instanceof Error ? error.message : "Arrow job not found",
				plugin: this.name
			});
		}
	}
	/**
	* Handle SQL query execution requests.
	* When called via asUser(req), uses the user's Databricks credentials.
	*/
	async _handleQueryRoute(req, res) {
		const { query_key } = req.params;
		const { parameters, format = "JSON" } = req.body;
		logger.debug(req, "Executing query: %s (format=%s)", query_key, format);
		logger.event(req)?.setComponent("analytics", "executeQuery").setContext("analytics", {
			query_key,
			format,
			parameter_count: parameters ? Object.keys(parameters).length : 0,
			plugin: this.name
		});
		if (!query_key) {
			res.status(400).json({ error: "query_key is required" });
			return;
		}
		const queryResult = await this.app.getAppQuery(query_key, req, this.devFileReader);
		if (!queryResult) {
			res.status(404).json({ error: "Query not found" });
			return;
		}
		const { query, isAsUser } = queryResult;
		const executor = isAsUser ? this.asUser(req) : this;
		const userKey = getCurrentUserId();
		const executorKey = isAsUser ? userKey : "global";
		const queryParameters = format === "ARROW" ? {
			formatParameters: {
				disposition: "EXTERNAL_LINKS",
				format: "ARROW_STREAM"
			},
			type: "arrow"
		} : { type: "result" };
		const hashedQuery = this.queryProcessor.hashQuery(query);
		const streamExecutionSettings = { default: {
			...queryDefaults,
			cache: {
				...queryDefaults.cache,
				cacheKey: [
					"analytics:query",
					query_key,
					JSON.stringify(parameters),
					JSON.stringify(format),
					hashedQuery,
					executorKey
				]
			}
		} };
		await executor.executeStream(res, async (signal) => {
			const processedParams = await this.queryProcessor.processQueryParams(query, parameters);
			const result = await executor.query(query, processedParams, queryParameters.formatParameters, signal);
			return {
				type: queryParameters.type,
				...result
			};
		}, streamExecutionSettings, executorKey);
	}
	/**
	* Execute a SQL query using the current execution context.
	*
	* When called directly: uses service principal credentials.
	* When called via asUser(req).query(...): uses user's credentials.
	*
	* @example
	* ```typescript
	* // Service principal execution
	* const result = await analytics.query("SELECT * FROM table")
	*
	* // User context execution (in route handler)
	* const result = await this.asUser(req).query("SELECT * FROM table")
	* ```
	*/
	async query(query, parameters, formatParameters, signal) {
		const workspaceClient = getWorkspaceClient();
		const warehouseId = await getWarehouseId();
		const { statement, parameters: sqlParameters } = this.queryProcessor.convertToSQLParameters(query, parameters);
		return (await this.SQLClient.executeStatement(workspaceClient, {
			statement,
			warehouse_id: warehouseId,
			parameters: sqlParameters,
			...formatParameters
		}, signal)).result;
	}
	/**
	* Get Arrow-formatted data for a completed query job.
	*/
	async getArrowData(workspaceClient, jobId, signal) {
		return await this.SQLClient.getArrowData(workspaceClient, jobId, signal);
	}
	async shutdown() {
		this.streamManager.abortAll();
	}
	/**
	* Returns the public exports for the analytics plugin.
	* Note: `asUser()` is automatically added by AppKit.
	*/
	exports() {
		return { query: this.query };
	}
};
/**
* @internal
*/
const analytics = toPlugin(AnalyticsPlugin, "analytics");

//#endregion
export { AnalyticsPlugin, analytics };
