// @databricks/appkit-ui/dist/shared/src/sql/helpers.js
//#region ../shared/src/sql/helpers.ts
/**
* SQL helper namespace
*/
const sql = {
	date(value) {
		let dateValue = "";
		if (value instanceof Date) dateValue = value.toISOString().split("T")[0];
		else if (typeof value === "string") {
			if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) throw new Error(`sql.date() expects Date or ISO date string (YYYY-MM-DD format), got: ${value}`);
			dateValue = value;
		} else throw new Error(`sql.date() expects Date or ISO date string (YYYY-MM-DD format), got: ${typeof value}`);
		return {
			__sql_type: "DATE",
			value: dateValue
		};
	},
	timestamp(value) {
		let timestampValue = "";
		if (value instanceof Date) timestampValue = value.toISOString().replace(/\.000(Z|[+-])/, "$1");
		else if (typeof value === "string") {
			if (!/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{1,9})?(Z|[+-]\d{2}:\d{2})?$/.test(value)) throw new Error(`sql.timestamp() expects ISO timestamp string (YYYY-MM-DDTHH:MM:SS.mmmZ or YYYY-MM-DDTHH:MM:SS.mmm+HH:MM), got: ${value}`);
			timestampValue = value;
		} else if (typeof value === "number") timestampValue = new Date(value > 0xe8d4a51000 ? value : value * 1e3).toISOString().replace(/\.000(Z|[+-])/, "$1");
		else throw new Error(`sql.timestamp() expects Date, ISO timestamp string, or Unix timestamp number, got: ${typeof value}`);
		return {
			__sql_type: "TIMESTAMP",
			value: timestampValue
		};
	},
	number(value) {
		let numValue = "";
		if (typeof value === "number") numValue = value.toString();
		else if (typeof value === "string") {
			if (value === "" || Number.isNaN(Number(value))) throw new Error(`sql.number() expects number or numeric string, got: ${value === "" ? "empty string" : value}`);
			numValue = value;
		} else throw new Error(`sql.number() expects number or numeric string, got: ${typeof value}`);
		return {
			__sql_type: "NUMERIC",
			value: numValue
		};
	},
	string(value) {
		if (typeof value !== "string" && typeof value !== "number" && typeof value !== "boolean") throw new Error(`sql.string() expects string or number or boolean, got: ${typeof value}`);
		let stringValue = "";
		if (typeof value === "string") stringValue = value;
		else stringValue = value.toString();
		return {
			__sql_type: "STRING",
			value: stringValue
		};
	},
	boolean(value) {
		if (typeof value !== "boolean" && typeof value !== "string" && typeof value !== "number") throw new Error(`sql.boolean() expects boolean or string (true or false) or number (1 or 0), got: ${typeof value}`);
		let booleanValue = "";
		if (typeof value === "boolean") booleanValue = value.toString();
		else if (typeof value === "number") {
			if (value !== 1 && value !== 0) throw new Error(`sql.boolean() expects boolean or string (true or false) or number (1 or 0), got: ${value}`);
			booleanValue = value === 1 ? "true" : "false";
		} else if (typeof value === "string") {
			if (value !== "true" && value !== "false") throw new Error(`sql.boolean() expects boolean or string (true or false) or number (1 or 0), got: ${value}`);
			booleanValue = value;
		}
		return {
			__sql_type: "BOOLEAN",
			value: booleanValue
		};
	},
	binary(value) {
		let hexValue = "";
		if (value instanceof Uint8Array) hexValue = Array.from(value).map((b) => b.toString(16).padStart(2, "0").toUpperCase()).join("");
		else if (value instanceof ArrayBuffer) hexValue = Array.from(new Uint8Array(value)).map((b) => b.toString(16).padStart(2, "0").toUpperCase()).join("");
		else if (typeof value === "string") {
			if (!/^[0-9A-Fa-f]*$/.test(value)) throw new Error(`sql.binary() expects Uint8Array, ArrayBuffer, or hex string, got invalid hex: ${value}`);
			hexValue = value.toUpperCase();
		} else throw new Error(`sql.binary() expects Uint8Array, ArrayBuffer, or hex string, got: ${typeof value}`);
		return {
			__sql_type: "STRING",
			value: hexValue
		};
	}
};
/**
* Type guard to check if a value is a SQL type marker
* @param value - Value to check
* @returns True if the value is a SQL type marker, false otherwise
* @example
* ```typescript
* const value = {
*   __sql_type: "DATE",
*   value: "2024-01-01",
* };
* const isSQLTypeMarker = isSQLTypeMarker(value);
* console.log(isSQLTypeMarker); // true
* ```
*/
function isSQLTypeMarker(value) {
	return value !== null && typeof value === "object" && "__sql_type" in value && "value" in value && typeof value.__sql_type === "string" && typeof value.value === "string";
}

//#endregion
export { isSQLTypeMarker, sql };
