// @databricks/appkit/dist/plugin/plugin.js
import { createLogger } from "../logging/logger.js";
import { AuthenticationError } from "../errors/authentication.js";
import { init_errors } from "../errors/index.js";
import { normalizeTelemetryOptions } from "../telemetry/config.js";
import { TelemetryManager } from "../telemetry/telemetry-manager.js";
import "../telemetry/index.js";
import { deepMerge } from "../utils/merge.js";
import { CacheManager } from "../cache/index.js";
import { ServiceContext } from "../context/service-context.js";
import { getCurrentUserId, runInUserContext } from "../context/execution-context.js";
import { init_context } from "../context/index.js";
import { AppManager } from "../app/index.js";
import { StreamManager } from "../stream/stream-manager.js";
import "../stream/index.js";
import { DevFileReader } from "./dev-reader.js";
import { CacheInterceptor } from "./interceptors/cache.js";
import { RetryInterceptor } from "./interceptors/retry.js";
import { TelemetryInterceptor } from "./interceptors/telemetry.js";
import { TimeoutInterceptor } from "./interceptors/timeout.js";

//#region src/plugin/plugin.ts
init_context();
init_errors();
const logger = createLogger("plugin");
/**
* Methods that should not be proxied by asUser().
* These are lifecycle/internal methods that don't make sense
* to execute in a user context.
*/
const EXCLUDED_FROM_PROXY = new Set([
	"setup",
	"shutdown",
	"injectRoutes",
	"getEndpoints",
	"abortActiveOperations",
	"asUser",
	"constructor"
]);
/**
* Base abstract class for creating AppKit plugins.
*
* All plugins must declare a static `manifest` property with their metadata
* and resource requirements. The manifest defines:
* - `required` resources: Always needed for the plugin to function
* - `optional` resources: May be needed depending on plugin configuration
*
* ## Static vs Runtime Resource Requirements
*
* The manifest is static and doesn't know the plugin's runtime configuration.
* For resources that become required based on config options, plugins can
* implement a static `getResourceRequirements(config)` method.
*
* At runtime, this method is called with the actual config to determine
* which "optional" resources should be treated as "required".
*
* @example Basic plugin with static requirements
* ```typescript
* import { Plugin, toPlugin, PluginManifest, ResourceType } from '@databricks/appkit';
*
* const myManifest: PluginManifest = {
*   name: 'myPlugin',
*   displayName: 'My Plugin',
*   description: 'Does something awesome',
*   resources: {
*     required: [
*       { type: ResourceType.SQL_WAREHOUSE, alias: 'warehouse', ... }
*     ],
*     optional: []
*   }
* };
*
* class MyPlugin extends Plugin<MyConfig> {
*   static manifest = myManifest;
*   name = 'myPlugin';
* }
* ```
*
* @example Plugin with config-dependent resources
* ```typescript
* interface MyConfig extends BasePluginConfig {
*   enableCaching?: boolean;
* }
*
* const myManifest: PluginManifest = {
*   name: 'myPlugin',
*   resources: {
*     required: [
*       { type: ResourceType.SQL_WAREHOUSE, alias: 'warehouse', ... }
*     ],
*     optional: [
*       // Database is optional in the static manifest
*       { type: ResourceType.DATABASE, alias: 'cache', description: 'Required if caching enabled', ... }
*     ]
*   }
* };
*
* class MyPlugin extends Plugin<MyConfig> {
*   static manifest = myManifest;
*   name = 'myPlugin';
*
*   // Runtime method: converts optional resources to required based on config
*   static getResourceRequirements(config: MyConfig) {
*     const resources = [];
*     if (config.enableCaching) {
*       // When caching is enabled, Database becomes required
*       resources.push({
*         type: ResourceType.DATABASE,
*         alias: 'cache',
*         resourceKey: 'database',
*         description: 'Cache storage for query results',
*         permission: 'CAN_CONNECT_AND_CREATE',
*         fields: {
*           instance_name: { env: 'DATABRICKS_CACHE_INSTANCE' },
*           database_name: { env: 'DATABRICKS_CACHE_DB' },
*         },
*         required: true  // Mark as required at runtime
*       });
*     }
*     return resources;
*   }
* }
* ```
*/
var Plugin = class {
	isReady = false;
	cache;
	app;
	devFileReader;
	streamManager;
	telemetry;
	/** Registered endpoints for this plugin */
	registeredEndpoints = {};
	/**
	* Plugin initialization phase.
	* - 'core': Initialized first (e.g., config plugins)
	* - 'normal': Initialized second (most plugins)
	* - 'deferred': Initialized last (e.g., server plugin)
	*/
	static phase = "normal";
	/**
	* Plugin name identifier.
	*/
	name;
	constructor(config) {
		this.config = config;
		this.name = config.name ?? "plugin";
		this.telemetry = TelemetryManager.getProvider(this.name, config.telemetry);
		this.streamManager = new StreamManager();
		this.cache = CacheManager.getInstanceSync();
		this.app = new AppManager();
		this.devFileReader = DevFileReader.getInstance();
		this.isReady = true;
	}
	injectRoutes(_) {}
	async setup() {}
	getEndpoints() {
		return this.registeredEndpoints;
	}
	abortActiveOperations() {
		this.streamManager.abortAll();
	}
	/**
	* Returns the public exports for this plugin.
	* Override this to define a custom public API.
	* By default, returns an empty object.
	*/
	exports() {
		return {};
	}
	/**
	* Execute operations using the user's identity from the request.
	* Returns a proxy of this plugin where all method calls execute
	* with the user's Databricks credentials instead of the service principal.
	*
	* @param req - The Express request containing the user token in headers
	* @returns A proxied plugin instance that executes as the user
	* @throws Error if user token is not available in request headers
	*/
	asUser(req) {
		const token = req.headers["x-forwarded-access-token"];
		const userId = req.headers["x-forwarded-user"];
		const isDev = process.env.NODE_ENV === "development";
		if (!token && isDev) {
			logger.warn("asUser() called without user token in development mode. Using service principal.");
			return this;
		}
		if (!token) throw AuthenticationError.missingToken("user token");
		if (!userId && !isDev) throw AuthenticationError.missingUserId();
		const effectiveUserId = userId || "dev-user";
		const userContext = ServiceContext.createUserContext(token, effectiveUserId);
		return this._createUserContextProxy(userContext);
	}
	/**
	* Creates a proxy that wraps method calls in a user context.
	* This allows all plugin methods to automatically use the user's
	* Databricks credentials.
	*/
	_createUserContextProxy(userContext) {
		return new Proxy(this, { get: (target, prop, receiver) => {
			const value = Reflect.get(target, prop, receiver);
			if (typeof value !== "function") return value;
			if (typeof prop === "string" && EXCLUDED_FROM_PROXY.has(prop)) return value;
			return (...args) => {
				return runInUserContext(userContext, () => value.apply(target, args));
			};
		} });
	}
	async executeStream(res, fn, options, userKey) {
		const { stream: streamConfig, default: defaultConfig, user: userConfig } = options;
		const executeConfig = this._buildExecutionConfig({
			default: defaultConfig,
			user: userConfig
		});
		const effectiveUserKey = userKey ?? getCurrentUserId();
		const self = this;
		const asyncWrapperFn = async function* (streamSignal) {
			const context = {
				signal: streamSignal,
				metadata: /* @__PURE__ */ new Map(),
				userKey: effectiveUserKey
			};
			const interceptors = self._buildInterceptors(executeConfig);
			const wrappedFn = async () => {
				return await fn(context.signal);
			};
			const result = await self._executeWithInterceptors(wrappedFn, interceptors, context);
			if (self._checkIfGenerator(result)) yield* result;
			else yield result;
		};
		await this.streamManager.stream(res, asyncWrapperFn, streamConfig);
	}
	async execute(fn, options, userKey) {
		const executeConfig = this._buildExecutionConfig(options);
		const interceptors = this._buildInterceptors(executeConfig);
		const effectiveUserKey = userKey ?? getCurrentUserId();
		const context = {
			metadata: /* @__PURE__ */ new Map(),
			userKey: effectiveUserKey
		};
		try {
			return await this._executeWithInterceptors(fn, interceptors, context);
		} catch (_error) {
			return;
		}
	}
	registerEndpoint(name, path) {
		this.registeredEndpoints[name] = path;
	}
	route(router, config) {
		const { name, method, path, handler } = config;
		router[method](path, handler);
		this.registerEndpoint(name, `/api/${this.name}${path}`);
	}
	_buildExecutionConfig(options) {
		const { default: methodDefaults, user: userOverride } = options;
		return deepMerge(deepMerge(methodDefaults, this.config), userOverride ?? {});
	}
	_buildInterceptors(options) {
		const interceptors = [];
		if (normalizeTelemetryOptions(this.config.telemetry).traces && (options.telemetryInterceptor?.enabled ?? true)) interceptors.push(new TelemetryInterceptor(this.telemetry, options.telemetryInterceptor));
		if (options.timeout && options.timeout > 0) interceptors.push(new TimeoutInterceptor(options.timeout));
		if (options.retry?.enabled && options.retry.attempts && options.retry.attempts > 1) interceptors.push(new RetryInterceptor(options.retry));
		if (options.cache?.enabled && options.cache.cacheKey?.length) interceptors.push(new CacheInterceptor(this.cache, options.cache));
		return interceptors;
	}
	async _executeWithInterceptors(fn, interceptors, context) {
		if (interceptors.length === 0) return fn(context.signal);
		let wrappedFn = () => fn(context.signal);
		for (const interceptor of interceptors) {
			const previousFn = wrappedFn;
			wrappedFn = () => interceptor.intercept(previousFn, context);
		}
		return wrappedFn();
	}
	_checkIfGenerator(result) {
		return result && typeof result === "object" && Symbol.asyncIterator in result;
	}
};

//#endregion
export { Plugin };
