// @databricks/appkit/dist/context/execution-context.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { ServiceContext, init_service_context } from "./service-context.js";
import { init_user_context, isUserContext } from "./user-context.js";
import { AsyncLocalStorage } from "node:async_hooks";

//#region src/context/execution-context.ts
/**
* Run a function in the context of a user.
* All calls within the function will have access to the user context.
*
* @param userContext - The user context to use
* @param fn - The function to run
* @returns The result of the function
*/
function runInUserContext(userContext, fn) {
	return executionContextStorage.run(userContext, fn);
}
/**
* Get the current execution context.
*
* - If running inside a user context (via asUser), returns the user context
* - Otherwise, returns the service context
*
* @throws Error if ServiceContext is not initialized
*/
function getExecutionContext() {
	const userContext = executionContextStorage.getStore();
	if (userContext) return userContext;
	return ServiceContext.get();
}
/**
* Get the current user ID for cache keying and telemetry.
*
* Returns the user ID if in user context, otherwise the service user ID.
*/
function getCurrentUserId() {
	const ctx = getExecutionContext();
	if (isUserContext(ctx)) return ctx.userId;
	return ctx.serviceUserId;
}
/**
* Get the WorkspaceClient for the current execution context.
*/
function getWorkspaceClient() {
	return getExecutionContext().client;
}
/**
* Get the warehouse ID promise.
*/
function getWarehouseId() {
	return getExecutionContext().warehouseId;
}
/**
* Get the workspace ID promise.
*/
function getWorkspaceId() {
	return getExecutionContext().workspaceId;
}
var executionContextStorage;
var init_execution_context = __esmMin((() => {
	init_service_context();
	init_user_context();
	executionContextStorage = new AsyncLocalStorage();
}));

//#endregion
init_execution_context();
export { getCurrentUserId, getExecutionContext, getWarehouseId, getWorkspaceClient, getWorkspaceId, init_execution_context, runInUserContext };
