// @databricks/appkit/dist/context/service-context.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { AuthenticationError } from "../errors/authentication.js";
import { ConfigurationError } from "../errors/configuration.js";
import { InitializationError } from "../errors/initialization.js";
import { init_errors } from "../errors/index.js";
import { name, version } from "../appkit/package.js";
import { WorkspaceClient } from "@databricks/sdk-experimental";
import { coerce } from "semver";

//#region src/context/service-context.ts
function getClientOptions() {
	const isDev = process.env.NODE_ENV === "development";
	return {
		product: name,
		productVersion: coerce(version)?.version ?? version,
		...isDev && { userAgentExtra: { mode: "dev" } }
	};
}
var ServiceContext;
var init_service_context = __esmMin((() => {
	init_errors();
	ServiceContext = class ServiceContext {
		static instance = null;
		static initPromise = null;
		/**
		* Initialize the service context. Should be called once at app startup.
		* Safe to call multiple times - will return the same instance.
		*
		* @param client - Optional pre-configured WorkspaceClient to use instead
		*   of creating one from environment credentials.
		*/
		static async initialize(client) {
			if (ServiceContext.instance) return ServiceContext.instance;
			if (ServiceContext.initPromise) return ServiceContext.initPromise;
			ServiceContext.initPromise = ServiceContext.createContext(client);
			ServiceContext.instance = await ServiceContext.initPromise;
			return ServiceContext.instance;
		}
		/**
		* Get the initialized service context.
		* @throws Error if not initialized
		*/
		static get() {
			if (!ServiceContext.instance) throw InitializationError.notInitialized("ServiceContext", "Call ServiceContext.initialize() first");
			return ServiceContext.instance;
		}
		/**
		* Check if the service context has been initialized.
		*/
		static isInitialized() {
			return ServiceContext.instance !== null;
		}
		/**
		* Create a user context from request headers.
		*
		* @param token - The user's access token from x-forwarded-access-token header
		* @param userId - The user's ID from x-forwarded-user header
		* @param userName - Optional user name
		* @throws Error if token is not provided
		*/
		static createUserContext(token, userId, userName) {
			if (!token) throw AuthenticationError.missingToken("user token");
			const host = process.env.DATABRICKS_HOST;
			if (!host) throw ConfigurationError.missingEnvVar("DATABRICKS_HOST");
			const serviceCtx = ServiceContext.get();
			return {
				client: new WorkspaceClient({
					token,
					host,
					authType: "pat"
				}, getClientOptions()),
				userId,
				userName,
				warehouseId: serviceCtx.warehouseId,
				workspaceId: serviceCtx.workspaceId,
				isUserContext: true
			};
		}
		/**
		* Get the client options for WorkspaceClient.
		* Exposed for testing purposes.
		*/
		static getClientOptions() {
			return getClientOptions();
		}
		static async createContext(client) {
			const wsClient = client ?? new WorkspaceClient({}, getClientOptions());
			const warehouseId = ServiceContext.getWarehouseId(wsClient);
			const workspaceId = ServiceContext.getWorkspaceId(wsClient);
			const currentUser = await wsClient.currentUser.me();
			if (!currentUser.id) throw ConfigurationError.resourceNotFound("Service user ID");
			return {
				client: wsClient,
				serviceUserId: currentUser.id,
				warehouseId,
				workspaceId
			};
		}
		static async getWorkspaceId(client) {
			if (process.env.DATABRICKS_WORKSPACE_ID) return process.env.DATABRICKS_WORKSPACE_ID;
			const response = await client.apiClient.request({
				path: "/api/2.0/preview/scim/v2/Me",
				method: "GET",
				headers: new Headers(),
				raw: false,
				query: {},
				responseHeaders: ["x-databricks-org-id"]
			});
			if (!response["x-databricks-org-id"]) throw ConfigurationError.resourceNotFound("Workspace ID");
			return response["x-databricks-org-id"];
		}
		static async getWarehouseId(client) {
			if (process.env.DATABRICKS_WAREHOUSE_ID) return process.env.DATABRICKS_WAREHOUSE_ID;
			if (process.env.NODE_ENV === "development") {
				const response = await client.apiClient.request({
					path: "/api/2.0/sql/warehouses",
					method: "GET",
					headers: new Headers(),
					raw: false,
					query: { skip_cannot_use: "true" }
				});
				const priorities = {
					RUNNING: 0,
					STOPPED: 1,
					STARTING: 2,
					STOPPING: 3,
					DELETED: 99,
					DELETING: 99
				};
				const warehouses = (response.warehouses || []).sort((a, b) => {
					return priorities[a.state] - priorities[b.state];
				});
				if (response.warehouses.length === 0) throw ConfigurationError.resourceNotFound("Warehouse ID", "Please configure the DATABRICKS_WAREHOUSE_ID environment variable");
				const firstWarehouse = warehouses[0];
				if (firstWarehouse.state === "DELETED" || firstWarehouse.state === "DELETING" || !firstWarehouse.id) throw ConfigurationError.resourceNotFound("Warehouse ID", "Please configure the DATABRICKS_WAREHOUSE_ID environment variable");
				return firstWarehouse.id;
			}
			throw ConfigurationError.resourceNotFound("Warehouse ID", "Please configure the DATABRICKS_WAREHOUSE_ID environment variable");
		}
		/**
		* Reset the service context. Only for testing purposes.
		*/
		static reset() {
			ServiceContext.instance = null;
			ServiceContext.initPromise = null;
		}
	};
}));

//#endregion
init_service_context();
export { ServiceContext, init_service_context };
