// @databricks/appkit/dist/core/appkit.js
import { TelemetryManager } from "../telemetry/telemetry-manager.js";
import "../telemetry/index.js";
import { CacheManager } from "../cache/index.js";
import { ServiceContext } from "../context/service-context.js";
import { init_context } from "../context/index.js";
import { ResourceRegistry } from "../registry/resource-registry.js";
import "../registry/index.js";

//#region src/core/appkit.ts
init_context();
var AppKit = class AppKit {
	#pluginInstances = {};
	#setupPromises = [];
	constructor(config) {
		const { plugins, ...globalConfig } = config;
		const pluginEntries = Object.entries(plugins);
		const corePlugins = pluginEntries.filter(([_, p]) => {
			return (p?.plugin?.phase ?? "normal") === "core";
		});
		const normalPlugins = pluginEntries.filter(([_, p]) => (p?.plugin?.phase ?? "normal") === "normal");
		const deferredPlugins = pluginEntries.filter(([_, p]) => (p?.plugin?.phase ?? "normal") === "deferred");
		for (const [name, pluginData] of corePlugins) if (pluginData) this.createAndRegisterPlugin(globalConfig, name, pluginData);
		for (const [name, pluginData] of normalPlugins) if (pluginData) this.createAndRegisterPlugin(globalConfig, name, pluginData);
		for (const [name, pluginData] of deferredPlugins) if (pluginData) this.createAndRegisterPlugin(globalConfig, name, pluginData, { plugins: this.#pluginInstances });
	}
	createAndRegisterPlugin(config, name, pluginData, extraData) {
		const { plugin: Plugin, config: pluginConfig } = pluginData;
		const pluginInstance = new Plugin({
			...config,
			...Plugin.DEFAULT_CONFIG,
			...pluginConfig,
			name,
			...extraData
		});
		this.#pluginInstances[name] = pluginInstance;
		this.#setupPromises.push(pluginInstance.setup());
		const self = this;
		Object.defineProperty(this, name, {
			get() {
				const plugin = self.#pluginInstances[name];
				return self.wrapWithAsUser(plugin);
			},
			enumerable: true
		});
	}
	/**
	* Binds all function properties in an exports object to the given context.
	*/
	bindExportMethods(exports, context) {
		for (const key in exports) if (Object.hasOwn(exports, key) && typeof exports[key] === "function") exports[key] = exports[key].bind(context);
	}
	/**
	* Wraps a plugin's exports with an `asUser` method that returns
	* a user-scoped version of the exports.
	*/
	wrapWithAsUser(plugin) {
		const pluginExports = plugin.exports?.() ?? {};
		this.bindExportMethods(pluginExports, plugin);
		if (typeof plugin.asUser !== "function") return pluginExports;
		return {
			...pluginExports,
			asUser: (req) => {
				const userPlugin = plugin.asUser(req);
				const userExports = userPlugin.exports?.() ?? {};
				this.bindExportMethods(userExports, userPlugin);
				return userExports;
			}
		};
	}
	static async _createApp(config = {}) {
		TelemetryManager.initialize(config?.telemetry);
		await CacheManager.getInstance(config?.cache);
		await ServiceContext.initialize(config?.client);
		const rawPlugins = config.plugins;
		const registry = new ResourceRegistry();
		registry.collectResources(rawPlugins);
		registry.enforceValidation();
		const instance = new AppKit({ plugins: AppKit.preparePlugins(rawPlugins) });
		await Promise.all(instance.#setupPromises);
		return instance;
	}
	static preparePlugins(plugins) {
		const result = {};
		for (const currentPlugin of plugins) result[currentPlugin.name] = {
			plugin: currentPlugin.plugin,
			config: currentPlugin.config
		};
		return result;
	}
};
/**
* Bootstraps AppKit with the provided configuration.
*/
async function createApp(config = {}) {
	return AppKit._createApp(config);
}

//#endregion
export { createApp };
