// @databricks/appkit/dist/registry/types.js
//#region src/registry/types.ts
/**
* Resource Registry Type System
*
* This module defines the type system for the AppKit Resource Registry,
* which enables plugins to declare their Databricks resource requirements
* in a machine-readable format.
*
* Resource types are exposed as first-class citizens with their specific
* permissions, making it simple for users to declare dependencies.
* Internal tooling handles conversion to Databricks app.yaml format.
*/
/**
* Supported resource types that plugins can depend on.
* Each type has its own set of valid permissions.
*/
let ResourceType = /* @__PURE__ */ function(ResourceType) {
	/** Secret scope for secure credential storage */
	ResourceType["SECRET"] = "secret";
	/** Databricks Job for scheduled or triggered workflows */
	ResourceType["JOB"] = "job";
	/** Databricks SQL Warehouse for query execution */
	ResourceType["SQL_WAREHOUSE"] = "sql_warehouse";
	/** Model serving endpoint for ML inference */
	ResourceType["SERVING_ENDPOINT"] = "serving_endpoint";
	/** Unity Catalog Volume for file storage */
	ResourceType["VOLUME"] = "volume";
	/** Vector Search Index for similarity search */
	ResourceType["VECTOR_SEARCH_INDEX"] = "vector_search_index";
	/** Unity Catalog Function */
	ResourceType["UC_FUNCTION"] = "uc_function";
	/** Unity Catalog Connection for external data sources */
	ResourceType["UC_CONNECTION"] = "uc_connection";
	/** Database (Lakebase) for persistent storage */
	ResourceType["DATABASE"] = "database";
	/** Genie Space for AI assistant */
	ResourceType["GENIE_SPACE"] = "genie_space";
	/** MLflow Experiment for ML tracking */
	ResourceType["EXPERIMENT"] = "experiment";
	/** Databricks App dependency */
	ResourceType["APP"] = "app";
	return ResourceType;
}({});
/**
* Permission hierarchy per resource type (weakest to strongest).
* Used to compare permissions when merging; higher index = more permissive.
* Unknown permissions are treated as less than any known permission.
*/
const PERMISSION_HIERARCHY_BY_TYPE = {
	[ResourceType.SECRET]: [
		"READ",
		"WRITE",
		"MANAGE"
	],
	[ResourceType.JOB]: [
		"CAN_VIEW",
		"CAN_MANAGE_RUN",
		"CAN_MANAGE"
	],
	[ResourceType.SQL_WAREHOUSE]: ["CAN_USE", "CAN_MANAGE"],
	[ResourceType.SERVING_ENDPOINT]: [
		"CAN_VIEW",
		"CAN_QUERY",
		"CAN_MANAGE"
	],
	[ResourceType.VOLUME]: ["READ_VOLUME", "WRITE_VOLUME"],
	[ResourceType.VECTOR_SEARCH_INDEX]: ["SELECT"],
	[ResourceType.UC_FUNCTION]: ["EXECUTE"],
	[ResourceType.UC_CONNECTION]: ["USE_CONNECTION"],
	[ResourceType.DATABASE]: ["CAN_CONNECT_AND_CREATE"],
	[ResourceType.GENIE_SPACE]: [
		"CAN_VIEW",
		"CAN_RUN",
		"CAN_EDIT",
		"CAN_MANAGE"
	],
	[ResourceType.EXPERIMENT]: [
		"CAN_READ",
		"CAN_EDIT",
		"CAN_MANAGE"
	],
	[ResourceType.APP]: ["CAN_USE"]
};
/** Set of valid permissions per type (for validation). */
const PERMISSIONS_BY_TYPE = PERMISSION_HIERARCHY_BY_TYPE;

//#endregion
export { PERMISSIONS_BY_TYPE, PERMISSION_HIERARCHY_BY_TYPE, ResourceType };
