// @databricks/appkit/dist/cache/index.js
import { createLogger } from "../logging/logger.js";
import { createLakebasePool } from "../connectors/lakebase/index.js";
import { AppKitError } from "../errors/base.js";
import { ExecutionError } from "../errors/execution.js";
import { InitializationError } from "../errors/initialization.js";
import { init_errors } from "../errors/index.js";
import { TelemetryManager } from "../telemetry/telemetry-manager.js";
import { SpanStatusCode } from "../telemetry/index.js";
import { deepMerge } from "../utils/merge.js";
import { cacheDefaults } from "./defaults.js";
import { InMemoryStorage } from "./storage/memory.js";
import { PersistentStorage } from "./storage/persistent.js";
import "./storage/index.js";
import { createHash } from "node:crypto";
import { WorkspaceClient } from "@databricks/sdk-experimental";

//#region src/cache/index.ts
init_errors();
const logger = createLogger("cache");
/**
* Cache manager class to handle cache operations.
* Can be used with in-memory storage or persistent storage (Lakebase).
*
* The cache is automatically initialized by AppKit. Use `getInstanceSync()` to access
* the singleton instance after initialization.
*
* @internal
* @example
* ```typescript
* const cache = CacheManager.getInstanceSync();
* const result = await cache.getOrExecute(["users", userId], () => fetchUser(userId), userKey);
* ```
*/
var CacheManager = class CacheManager {
	static MIN_CLEANUP_INTERVAL_MS = 6e4;
	name = "cache-manager";
	static instance = null;
	static initPromise = null;
	storage;
	config;
	inFlightRequests;
	cleanupInProgress;
	lastCleanupAttempt;
	telemetry;
	telemetryMetrics;
	constructor(storage, config) {
		this.storage = storage;
		this.config = config;
		this.inFlightRequests = /* @__PURE__ */ new Map();
		this.cleanupInProgress = false;
		this.lastCleanupAttempt = 0;
		this.telemetry = TelemetryManager.getProvider(this.name, this.config.telemetry);
		this.telemetryMetrics = {
			cacheHitCount: this.telemetry.getMeter().createCounter("cache.hit", {
				description: "Total number of cache hits",
				unit: "1"
			}),
			cacheMissCount: this.telemetry.getMeter().createCounter("cache.miss", {
				description: "Total number of cache misses",
				unit: "1"
			})
		};
	}
	/**
	* Get the singleton instance of the cache manager (sync version).
	*
	* Throws if not initialized - ensure AppKit.create() has completed first.
	* @returns CacheManager instance
	*/
	static getInstanceSync() {
		if (!CacheManager.instance) throw InitializationError.notInitialized("CacheManager", "Ensure AppKit.create() has completed before accessing the cache");
		return CacheManager.instance;
	}
	/**
	* Initialize and get the singleton instance of the cache manager.
	* Called internally by AppKit - prefer `getInstanceSync()` for plugin access.
	* @param userConfig - User configuration for the cache manager
	* @returns CacheManager instance
	* @internal
	*/
	static async getInstance(userConfig) {
		if (CacheManager.instance) return CacheManager.instance;
		if (!CacheManager.initPromise) CacheManager.initPromise = CacheManager.create(userConfig).then((instance) => {
			CacheManager.instance = instance;
			return instance;
		});
		return CacheManager.initPromise;
	}
	/**
	* Create a new cache manager instance
	*
	* Storage selection logic:
	* 1. If `storage` provided and healthy → use provided storage
	* 2. If `storage` provided but unhealthy → fallback to InMemory (or disable if strictPersistence)
	* 3. If no `storage` provided and Lakebase available → use Lakebase
	* 4. If no `storage` provided and Lakebase unavailable → fallback to InMemory (or disable if strictPersistence)
	*
	* @param userConfig - User configuration for the cache manager
	* @returns CacheManager instance
	*/
	static async create(userConfig) {
		const config = deepMerge(cacheDefaults, userConfig);
		if (config.storage) {
			if (await config.storage.healthCheck()) return new CacheManager(config.storage, config);
			if (config.strictPersistence) {
				const disabledConfig = {
					...config,
					enabled: false
				};
				return new CacheManager(new InMemoryStorage(disabledConfig), disabledConfig);
			}
			return new CacheManager(new InMemoryStorage(config), config);
		}
		try {
			const pool = createLakebasePool({ workspaceClient: new WorkspaceClient({}) });
			const persistentStorage = new PersistentStorage(config, pool);
			if (await persistentStorage.healthCheck()) {
				await persistentStorage.initialize();
				return new CacheManager(persistentStorage, config);
			}
			await pool.end();
		} catch {}
		if (config.strictPersistence) {
			const disabledConfig = {
				...config,
				enabled: false
			};
			return new CacheManager(new InMemoryStorage(disabledConfig), disabledConfig);
		}
		return new CacheManager(new InMemoryStorage(config), config);
	}
	/**
	* Get or execute a function and cache the result
	* @param key - Cache key
	* @param fn - Function to execute
	* @param userKey - User key
	* @param options - Options for the cache
	* @returns Promise of the result
	*/
	async getOrExecute(key, fn, userKey, options) {
		if (!this.config.enabled) return fn();
		const cacheKey = this.generateKey(key, userKey);
		return this.telemetry.startActiveSpan("cache.getOrExecute", { attributes: {
			"cache.key": cacheKey,
			"cache.enabled": this.config.enabled,
			"cache.persistent": this.storage.isPersistent()
		} }, async (span) => {
			try {
				const cached = await this.storage.get(cacheKey);
				if (cached !== null) {
					span.setAttribute("cache.hit", true);
					span.setStatus({ code: SpanStatusCode.OK });
					this.telemetryMetrics.cacheHitCount.add(1, { "cache.key": cacheKey });
					logger.event()?.setExecution({
						cache_hit: true,
						cache_key: cacheKey
					});
					return cached.value;
				}
				const inFlight = this.inFlightRequests.get(cacheKey);
				if (inFlight) {
					span.setAttribute("cache.hit", true);
					span.setAttribute("cache.deduplication", true);
					span.addEvent("cache.deduplication_used", { "cache.key": cacheKey });
					span.setStatus({ code: SpanStatusCode.OK });
					this.telemetryMetrics.cacheHitCount.add(1, {
						"cache.key": cacheKey,
						"cache.deduplication": "true"
					});
					logger.event()?.setExecution({
						cache_hit: true,
						cache_key: cacheKey,
						cache_deduplication: true
					});
					span.end();
					return inFlight;
				}
				span.setAttribute("cache.hit", false);
				span.addEvent("cache.miss", { "cache.key": cacheKey });
				this.telemetryMetrics.cacheMissCount.add(1, { "cache.key": cacheKey });
				logger.event()?.setExecution({
					cache_hit: false,
					cache_key: cacheKey
				});
				const promise = fn().then(async (result) => {
					await this.set(cacheKey, result, options);
					span.addEvent("cache.value_stored", {
						"cache.key": cacheKey,
						"cache.ttl": options?.ttl ?? this.config.ttl ?? 3600
					});
					return result;
				}).catch((error) => {
					span.recordException(error);
					span.setStatus({ code: SpanStatusCode.ERROR });
					if (error instanceof AppKitError) throw error;
					throw ExecutionError.statementFailed(error instanceof Error ? error.message : String(error));
				}).finally(() => {
					this.inFlightRequests.delete(cacheKey);
				});
				this.inFlightRequests.set(cacheKey, promise);
				const result = await promise;
				span.setStatus({ code: SpanStatusCode.OK });
				return result;
			} catch (error) {
				span.recordException(error);
				span.setStatus({ code: SpanStatusCode.ERROR });
				throw error;
			} finally {
				span.end();
			}
		}, {
			name: this.name,
			includePrefix: true
		});
	}
	/**
	* Get a cached value
	* @param key - Cache key
	* @returns Promise of the value or null if not found or expired
	*/
	async get(key) {
		if (!this.config.enabled) return null;
		this.maybeCleanup();
		const entry = await this.storage.get(key);
		if (!entry) return null;
		if (Date.now() > entry.expiry) {
			await this.storage.delete(key);
			return null;
		}
		return entry.value;
	}
	/** Probabilistically trigger cleanup of expired entries (fire-and-forget) */
	maybeCleanup() {
		if (this.cleanupInProgress) return;
		if (!this.storage.isPersistent()) return;
		const now = Date.now();
		if (now - this.lastCleanupAttempt < CacheManager.MIN_CLEANUP_INTERVAL_MS) return;
		const probability = this.config.cleanupProbability ?? .01;
		if (Math.random() > probability) return;
		this.lastCleanupAttempt = now;
		this.cleanupInProgress = true;
		this.storage.cleanupExpired().catch((error) => {
			logger.debug("Error cleaning up expired entries: %O", error);
		}).finally(() => {
			this.cleanupInProgress = false;
		});
	}
	/**
	* Set a value in the cache
	* @param key - Cache key
	* @param value - Value to set
	* @param options - Options for the cache
	* @returns Promise of the result
	*/
	async set(key, value, options) {
		if (!this.config.enabled) return;
		const ttl = options?.ttl ?? this.config.ttl ?? 3600;
		const expiryTime = Date.now() + ttl * 1e3;
		await this.storage.set(key, {
			value,
			expiry: expiryTime
		});
	}
	/**
	* Delete a value from the cache
	* @param key - Cache key
	* @returns Promise of the result
	*/
	async delete(key) {
		if (!this.config.enabled) return;
		await this.storage.delete(key);
	}
	/** Clear the cache */
	async clear() {
		await this.storage.clear();
		this.inFlightRequests.clear();
	}
	/**
	* Check if a value exists in the cache
	* @param key - Cache key
	* @returns Promise of true if the value exists, false otherwise
	*/
	async has(key) {
		if (!this.config.enabled) return false;
		const entry = await this.storage.get(key);
		if (!entry) return false;
		if (Date.now() > entry.expiry) {
			await this.storage.delete(key);
			return false;
		}
		return true;
	}
	/**
	* Generate a cache key
	* @param parts - Parts of the key
	* @param userKey - User key
	* @returns Cache key
	*/
	generateKey(parts, userKey) {
		const allParts = [userKey, ...parts];
		const serialized = JSON.stringify(allParts);
		return createHash("sha256").update(serialized).digest("hex");
	}
	/** Close the cache */
	async close() {
		await this.storage.close();
	}
	/**
	* Check if the storage is healthy
	* @returns Promise of true if the storage is healthy, false otherwise
	*/
	async isStorageHealthy() {
		return this.storage.healthCheck();
	}
};

//#endregion
export { CacheManager };
