// @databricks/appkit/dist/index.js
import { isSQLTypeMarker, sql } from "./shared/src/sql/helpers.js";
import { RequestedClaimsPermissionSet, createLakebasePool, generateDatabaseCredential, getLakebaseOrmConfig, getLakebasePgConfig, getWorkspaceClient } from "./connectors/lakebase/index.js";
import { AppKitError } from "./errors/base.js";
import { AuthenticationError } from "./errors/authentication.js";
import { ConfigurationError } from "./errors/configuration.js";
import { ConnectionError } from "./errors/connection.js";
import { ExecutionError } from "./errors/execution.js";
import { InitializationError } from "./errors/initialization.js";
import { ServerError } from "./errors/server.js";
import { TunnelError } from "./errors/tunnel.js";
import { ValidationError } from "./errors/validation.js";
import { init_errors } from "./errors/index.js";
import { SeverityNumber, SpanStatusCode } from "./telemetry/index.js";
import { CacheManager } from "./cache/index.js";
import { getExecutionContext } from "./context/execution-context.js";
import { init_context } from "./context/index.js";
import { ResourceType } from "./registry/types.js";
import { getPluginManifest, getResourceRequirements } from "./registry/manifest-loader.js";
import { ResourceRegistry } from "./registry/resource-registry.js";
import "./registry/index.js";
import { createApp } from "./core/appkit.js";
import "./core/index.js";
import { Plugin } from "./plugin/plugin.js";
import { toPlugin } from "./plugin/to-plugin.js";
import "./plugin/index.js";
import { analytics } from "./plugins/analytics/analytics.js";
import { appKitTypesPlugin } from "./type-generator/vite-plugin.js";
import { server } from "./plugins/server/index.js";
import "./plugins/index.js";

//#region src/index.ts
init_context();
init_errors();

//#endregion
export { AppKitError, AuthenticationError, CacheManager, ConfigurationError, ConnectionError, ExecutionError, InitializationError, Plugin, RequestedClaimsPermissionSet, ResourceRegistry, ResourceType, ServerError, SeverityNumber, SpanStatusCode, TunnelError, ValidationError, analytics, appKitTypesPlugin, createApp, createLakebasePool, generateDatabaseCredential, getExecutionContext, getLakebaseOrmConfig, getLakebasePgConfig, getPluginManifest, getResourceRequirements, getWorkspaceClient, isSQLTypeMarker, server, sql, toPlugin };
