// @databricks/appkit/dist/type-generator/cache.js
import crypto from "node:crypto";
import path from "node:path";
import fs from "node:fs";

//#region src/type-generator/cache.ts
const CACHE_VERSION = "1";
const CACHE_FILE = ".appkit-types-cache.json";
const CACHE_DIR = path.join(process.cwd(), "node_modules", ".databricks", "appkit");
/**
* Hash the SQL query
* Uses MD5 to hash the SQL query
* @param sql - the SQL query to hash
* @returns - the hash of the SQL query
*/
function hashSQL(sql) {
	return crypto.createHash("md5").update(sql).digest("hex");
}
/**
* Load the cache from the file system
* If the cache is not found, run the query explain
* @returns - the cache
*/
function loadCache() {
	const cachePath = path.join(CACHE_DIR, CACHE_FILE);
	try {
		if (!fs.existsSync(CACHE_DIR)) fs.mkdirSync(CACHE_DIR, { recursive: true });
		if (fs.existsSync(cachePath)) {
			const cache = JSON.parse(fs.readFileSync(cachePath, "utf8"));
			if (cache.version === CACHE_VERSION) return cache;
		}
	} catch {}
	return {
		version: CACHE_VERSION,
		queries: {}
	};
}
/**
* Save the cache to the file system
* The cache is saved as a JSON file, it is used to avoid running the query explain multiple times
* @param cache - cache object to save
*/
function saveCache(cache) {
	const cachePath = path.join(CACHE_DIR, CACHE_FILE);
	fs.writeFileSync(cachePath, JSON.stringify(cache, null, 2), "utf8");
}

//#endregion
export { CACHE_VERSION, hashSQL, loadCache, saveCache };
