// @databricks/appkit/dist/type-generator/query-registry.js
import { createLogger } from "../logging/logger.js";
import { CACHE_VERSION, hashSQL, loadCache, saveCache } from "./cache.js";
import { Spinner } from "./spinner.js";
import { sqlTypeToHelper, sqlTypeToMarker } from "./types.js";
import { WorkspaceClient } from "@databricks/sdk-experimental";
import path from "node:path";
import fs from "node:fs";

//#region src/type-generator/query-registry.ts
const logger = createLogger("type-generator:query-registry");
/**
* Extract parameters from a SQL query
* @param sql - the SQL query to extract parameters from
* @returns an array of parameter names
*/
function extractParameters(sql) {
	const matches = sql.matchAll(/:([a-zA-Z_]\w*)/g);
	const params = /* @__PURE__ */ new Set();
	for (const match of matches) params.add(match[1]);
	return Array.from(params);
}
const SERVER_INJECTED_PARAMS = ["workspaceId"];
function convertToQueryType(result, sql, queryName) {
	const columns = (result.result?.data_array || []).map((row) => ({
		name: row[0] || "",
		type_name: row[1]?.toUpperCase() || "STRING",
		comment: row[2] || void 0
	}));
	const params = extractParameters(sql).filter((p) => !SERVER_INJECTED_PARAMS.includes(p));
	const paramTypes = extractParameterTypes(sql);
	return `{
    name: "${queryName}";
    parameters: ${params.length > 0 ? `{\n      ${params.map((p) => {
		const sqlType = paramTypes[p];
		const markerType = sqlType ? sqlTypeToMarker[sqlType] : "SQLTypeMarker";
		const helper = sqlType ? sqlTypeToHelper[sqlType] : "sql.*()";
		return `/** ${sqlType || "any"} - use ${helper} */\n      ${p}: ${markerType}`;
	}).join(";\n      ")};\n    }` : "Record<string, never>"};
    result: Array<{
      ${columns.map((column) => {
		const mappedType = typeMap[normalizeTypeName(column.type_name)] || "unknown";
		const name = /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(column.name) ? column.name : `"${column.name}"`;
		return `${column.comment ? `/** ${column.comment} */\n      ` : `/** @sqlType ${column.type_name} */\n      `}${name}: ${mappedType}`;
	}).join(";\n      ")};
    }>;
  }`;
}
function extractParameterTypes(sql) {
	const paramTypes = {};
	const matches = sql.matchAll(/--\s*@param\s+(\w+)\s+(STRING|NUMERIC|BOOLEAN|DATE|TIMESTAMP|BINARY)/gi);
	for (const match of matches) {
		const [, paramName, paramType] = match;
		paramTypes[paramName] = paramType.toUpperCase();
	}
	return paramTypes;
}
/**
* Generate query schemas from a folder of SQL files
* It uses DESCRIBE QUERY to get the schema without executing the query
* @param queryFolder - the folder containing the SQL files
* @param warehouseId - the warehouse id to use for schema analysis
* @param options - options for the query generation
* @param options.noCache - if true, skip the cache and regenerate all types
* @returns an array of query schemas
*/
async function generateQueriesFromDescribe(queryFolder, warehouseId, options = {}) {
	const { noCache = false } = options;
	const queryFiles = fs.readdirSync(queryFolder).filter((file) => file.endsWith(".sql"));
	logger.debug("Found %d SQL queries", queryFiles.length);
	const cache = noCache ? {
		version: CACHE_VERSION,
		queries: {}
	} : loadCache();
	const client = new WorkspaceClient({});
	const querySchemas = [];
	const failedQueries = [];
	const spinner = new Spinner();
	for (let i = 0; i < queryFiles.length; i++) {
		const file = queryFiles[i];
		const queryName = normalizeQueryName(path.basename(file, ".sql"));
		const sql = fs.readFileSync(path.join(queryFolder, file), "utf8");
		const sqlHash = hashSQL(sql);
		const cached = cache.queries[queryName];
		if (cached && cached.hash === sqlHash) {
			querySchemas.push({
				name: queryName,
				type: cached.type
			});
			spinner.start(`Processing ${queryName} (${i + 1}/${queryFiles.length})`);
			spinner.stop(`✓ ${queryName} (cached)`);
			continue;
		}
		spinner.start(`Processing ${queryName} (${i + 1}/${queryFiles.length})`);
		const cleanedSql = sql.replace(/:([a-zA-Z_]\w*)/g, "''").trim().replace(/;\s*$/, "");
		try {
			const result = await client.statementExecution.executeStatement({
				statement: `DESCRIBE QUERY ${cleanedSql}`,
				warehouse_id: warehouseId
			});
			if (result.status.state === "FAILED") {
				const sqlError = result.status.error?.message || "Query execution failed";
				spinner.stop(`✗ ${queryName} - failed`);
				spinner.printDetail(`SQL Error: ${sqlError}`);
				spinner.printDetail(`Query: ${cleanedSql.slice(0, 200)}`);
				failedQueries.push({
					name: queryName,
					error: sqlError
				});
				continue;
			}
			const type = convertToQueryType(result, sql, queryName);
			querySchemas.push({
				name: queryName,
				type
			});
			cache.queries[queryName] = {
				hash: sqlHash,
				type
			};
			spinner.stop(`✓ ${queryName}`);
		} catch (error) {
			const errorMessage = error instanceof Error ? error.message : "Unknown error";
			spinner.stop(`✗ ${queryName} - ${errorMessage}`);
			failedQueries.push({
				name: queryName,
				error: errorMessage
			});
		}
	}
	saveCache(cache);
	if (failedQueries.length > 0) logger.debug("Warning: %d queries failed", failedQueries.length);
	return querySchemas;
}
/**
* Normalize query name by removing the .obo extension
* @param queryName - the query name to normalize
* @returns the normalized query name
*/
function normalizeQueryName(fileName) {
	return fileName.replace(/\.obo$/, "");
}
/**
* Normalize SQL type name by removing parameters/generics
* Examples:
*   DECIMAL(38,6) -> DECIMAL
*   ARRAY<STRING> -> ARRAY
*   MAP<STRING,INT> -> MAP
*   STRUCT<name:STRING> -> STRUCT
*   INTERVAL DAY TO SECOND -> INTERVAL
*   GEOGRAPHY(4326) -> GEOGRAPHY
*/
function normalizeTypeName(typeName) {
	return typeName.replace(/\(.*\)$/, "").replace(/<.*>$/, "").split(" ")[0];
}
/** Type Map for Databricks data types to JavaScript types */
const typeMap = {
	STRING: "string",
	BINARY: "string",
	BOOLEAN: "boolean",
	TINYINT: "number",
	SMALLINT: "number",
	INT: "number",
	BIGINT: "number",
	FLOAT: "number",
	DOUBLE: "number",
	DECIMAL: "number",
	DATE: "string",
	TIMESTAMP: "string",
	TIMESTAMP_NTZ: "string",
	INTERVAL: "string",
	ARRAY: "unknown[]",
	MAP: "Record<string, unknown>",
	STRUCT: "Record<string, unknown>",
	OBJECT: "Record<string, unknown>",
	VARIANT: "unknown",
	GEOGRAPHY: "unknown",
	GEOMETRY: "unknown",
	VOID: "null"
};

//#endregion
export { generateQueriesFromDescribe };
