// @databricks/appkit/dist/type-generator/types.js
//#region src/type-generator/types.ts
/**
* Map of SQL types to their corresponding marker types
* Used to convert SQL types to their corresponding marker types
*/
const sqlTypeToMarker = {
	STRING: "SQLStringMarker",
	BINARY: "SQLBinaryMarker",
	BOOLEAN: "SQLBooleanMarker",
	NUMERIC: "SQLNumberMarker",
	INT: "SQLNumberMarker",
	BIGINT: "SQLNumberMarker",
	TINYINT: "SQLNumberMarker",
	SMALLINT: "SQLNumberMarker",
	FLOAT: "SQLNumberMarker",
	DOUBLE: "SQLNumberMarker",
	DECIMAL: "SQLNumberMarker",
	DATE: "SQLDateMarker",
	TIMESTAMP: "SQLTimestampMarker",
	TIMESTAMP_NTZ: "SQLTimestampMarker"
};
/**
* Map of SQL types to their corresponding helper function names
* Used to generate JSDoc hints for parameters
*/
const sqlTypeToHelper = {
	STRING: "sql.string()",
	BINARY: "sql.binary()",
	BOOLEAN: "sql.boolean()",
	NUMERIC: "sql.number()",
	INT: "sql.number()",
	BIGINT: "sql.number()",
	TINYINT: "sql.number()",
	SMALLINT: "sql.number()",
	FLOAT: "sql.number()",
	DOUBLE: "sql.number()",
	DECIMAL: "sql.number()",
	DATE: "sql.date()",
	TIMESTAMP: "sql.timestamp()",
	TIMESTAMP_NTZ: "sql.timestamp()"
};

//#endregion
export { sqlTypeToHelper, sqlTypeToMarker };
