// @databricks/appkit/dist/type-generator/vite-plugin.js
import { createLogger } from "../logging/logger.js";
import { generateFromEntryPoint } from "./index.js";
import path from "node:path";
import fs from "node:fs";

//#region src/type-generator/vite-plugin.ts
const logger = createLogger("type-generator:vite-plugin");
/**
* Vite plugin to generate types for AppKit queries.
* Calls generateFromEntryPoint under the hood.
* @param options - Options to override default values.
* @returns Vite plugin to generate types for AppKit queries.
*/
function appKitTypesPlugin(options) {
	let root;
	let outFile;
	let watchFolders;
	async function generate() {
		try {
			const warehouseId = process.env.DATABRICKS_WAREHOUSE_ID || "";
			if (!warehouseId) {
				logger.debug("Warehouse ID not found. Skipping type generation.");
				return;
			}
			await generateFromEntryPoint({
				outFile,
				queryFolder: watchFolders[0],
				warehouseId,
				noCache: false
			});
		} catch (error) {
			if (process.env.NODE_ENV === "production") throw error;
			logger.error("Error generating types: %O", error);
		}
	}
	return {
		name: "appkit-types",
		apply() {
			if (!(process.env.DATABRICKS_WAREHOUSE_ID || "")) {
				logger.debug("Warehouse ID not found. Skipping type generation.");
				return false;
			}
			if (!fs.existsSync(path.join(process.cwd(), "config", "queries"))) return false;
			return true;
		},
		configResolved(config) {
			root = config.root;
			outFile = path.resolve(root, options?.outFile ?? "src/appKitTypes.d.ts");
			watchFolders = options?.watchFolders ?? [path.join(process.cwd(), "config", "queries")];
		},
		buildStart() {
			generate();
		},
		configureServer(server) {
			server.watcher.add(watchFolders);
			server.watcher.on("change", (changedFile) => {
				if (watchFolders.some((folder) => changedFile.startsWith(folder)) && changedFile.endsWith(".sql")) generate();
			});
		}
	};
}

//#endregion
export { appKitTypesPlugin };
