// @databricks/appkit/dist/plugin/interceptors/retry.js
import { createLogger } from "../../logging/logger.js";

//#region src/plugin/interceptors/retry.ts
const logger = createLogger("interceptors:retry");
var RetryInterceptor = class {
	attempts;
	initialDelay;
	maxDelay;
	constructor(config) {
		this.attempts = config.attempts ?? 3;
		this.initialDelay = config.initialDelay ?? 1e3;
		this.maxDelay = config.maxDelay ?? 3e4;
	}
	async intercept(fn, context) {
		let lastError;
		for (let attempt = 1; attempt <= this.attempts; attempt++) try {
			const result = await fn();
			if (attempt > 1) logger.event()?.setExecution({ retry_attempts: attempt - 1 });
			return result;
		} catch (error) {
			lastError = error;
			if (attempt === this.attempts) {
				logger.event()?.setExecution({ retry_attempts: attempt - 1 });
				throw error;
			}
			if (context.signal?.aborted) throw error;
			const delay = this.calculateDelay(attempt);
			await this.sleep(delay);
		}
		throw lastError;
	}
	calculateDelay(attempt) {
		const delay = this.initialDelay * 2 ** (attempt - 1);
		return Math.min(delay, this.maxDelay);
	}
	sleep(ms) {
		return new Promise((resolve) => setTimeout(resolve, ms));
	}
};

//#endregion
export { RetryInterceptor };
