// @databricks/appkit/dist/errors/authentication.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { AppKitError, init_base } from "./base.js";

//#region src/errors/authentication.ts
var AuthenticationError;
var init_authentication = __esmMin((() => {
	init_base();
	AuthenticationError = class AuthenticationError extends AppKitError {
		code = "AUTHENTICATION_ERROR";
		statusCode = 401;
		isRetryable = false;
		/**
		* Create an authentication error for missing token
		*/
		static missingToken(tokenType = "access token") {
			return new AuthenticationError(`Missing ${tokenType} in request headers`, { context: { tokenType } });
		}
		/**
		* Create an authentication error for missing user identity
		*/
		static missingUserId() {
			return new AuthenticationError("User ID not available in request headers. Ensure the request has the x-forwarded-user header.");
		}
		/**
		* Create an authentication error for credential generation failure
		*/
		static credentialsFailed(instance, cause) {
			return new AuthenticationError(`Failed to generate credentials for instance: ${instance}`, {
				cause,
				context: { instance }
			});
		}
		/**
		* Create an authentication error for failed user lookup
		*/
		static userLookupFailed(cause) {
			return new AuthenticationError("Failed to get current user from Databricks workspace", { cause });
		}
	};
}));

//#endregion
init_authentication();
export { AuthenticationError, init_authentication };
