// @databricks/appkit/dist/errors/base.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";

//#region src/errors/base.ts
var SENSITIVE_FIELD_PATTERNS, AppKitError;
var init_base = __esmMin((() => {
	SENSITIVE_FIELD_PATTERNS = [
		/token/i,
		/password/i,
		/secret/i,
		/credential/i,
		/auth/i,
		/key$/i,
		/apikey/i
	];
	AppKitError = class extends Error {
		/** Optional cause of the error */
		cause;
		/** Additional context for the error */
		context;
		constructor(message, options) {
			super(message);
			this.name = this.constructor.name;
			this.cause = options?.cause;
			this.context = options?.context;
			if (Error.captureStackTrace) Error.captureStackTrace(this, this.constructor);
		}
		/**
		* Convert error to JSON for logging/serialization.
		* Sensitive values in context are automatically redacted.
		*/
		toJSON() {
			return {
				name: this.name,
				code: this.code,
				message: this.message,
				statusCode: this.statusCode,
				isRetryable: this.isRetryable,
				context: this.sanitizeContext(this.context),
				cause: this.cause?.message,
				stack: this.stack
			};
		}
		/**
		* Create a human-readable string representation
		*/
		toString() {
			let str = `${this.name} [${this.code}]: ${this.message}`;
			if (this.cause) str += ` (caused by: ${this.cause.message})`;
			return str;
		}
		/**
		* Sanitize context by redacting sensitive field values
		*/
		sanitizeContext(context) {
			if (!context) return void 0;
			const sanitized = {};
			for (const [key, value] of Object.entries(context)) if (this.isSensitiveField(key)) sanitized[key] = "[REDACTED]";
			else if (typeof value === "object" && value !== null) sanitized[key] = Array.isArray(value) ? `[Array(${value.length})]` : "[Object]";
			else sanitized[key] = value;
			return sanitized;
		}
		/**
		* Check if a field name matches sensitive patterns
		*/
		isSensitiveField(fieldName) {
			return SENSITIVE_FIELD_PATTERNS.some((pattern) => pattern.test(fieldName));
		}
	};
}));

//#endregion
init_base();
export { AppKitError, init_base };
