// @databricks/appkit/dist/errors/configuration.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { AppKitError, init_base } from "./base.js";

//#region src/errors/configuration.ts
var ConfigurationError;
var init_configuration = __esmMin((() => {
	init_base();
	ConfigurationError = class ConfigurationError extends AppKitError {
		code = "CONFIGURATION_ERROR";
		statusCode = 500;
		isRetryable = false;
		/**
		* Create a configuration error for missing environment variable
		*/
		static missingEnvVar(varName) {
			return new ConfigurationError(`${varName} environment variable is required`, { context: { envVar: varName } });
		}
		/**
		* Create a configuration error for missing resource
		*/
		static resourceNotFound(resource, hint) {
			return new ConfigurationError(hint ? `${resource} not found. ${hint}` : `${resource} not found`, { context: { resource } });
		}
		/**
		* Create a configuration error for invalid connection config
		*/
		static invalidConnection(service, details) {
			return new ConfigurationError(details ? `${service} connection not configured. ${details}` : `${service} connection not configured`, { context: { service } });
		}
		/**
		* Create a configuration error for missing connection string parameter
		*/
		static missingConnectionParam(param) {
			return new ConfigurationError(`Connection string must include ${param} parameter`, { context: { parameter: param } });
		}
	};
}));

//#endregion
init_configuration();
export { ConfigurationError, init_configuration };
