// @databricks/appkit/dist/errors/connection.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { AppKitError, init_base } from "./base.js";

//#region src/errors/connection.ts
var ConnectionError;
var init_connection = __esmMin((() => {
	init_base();
	ConnectionError = class ConnectionError extends AppKitError {
		code = "CONNECTION_ERROR";
		statusCode = 503;
		isRetryable = true;
		/**
		* Create a connection error for query failure
		*/
		static queryFailed(cause) {
			return new ConnectionError("Query failed", { cause });
		}
		/**
		* Create a connection error for transaction failure
		*/
		static transactionFailed(cause) {
			return new ConnectionError("Transaction failed", { cause });
		}
		/**
		* Create a connection error for pool errors
		*/
		static poolError(operation, cause) {
			return new ConnectionError(`Connection pool error: ${operation}`, { cause });
		}
		/**
		* Create a connection error for API failures
		*/
		static apiFailure(service, cause) {
			return new ConnectionError(`No response received from ${service} API`, {
				cause,
				context: { service }
			});
		}
		/**
		* Create a connection error for client unavailable
		*/
		static clientUnavailable(clientType, hint) {
			return new ConnectionError(hint ? `${clientType} not available. ${hint}` : `${clientType} not available`, { context: { clientType } });
		}
	};
}));

//#endregion
init_connection();
export { ConnectionError, init_connection };
