// @databricks/appkit/dist/errors/initialization.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { AppKitError, init_base } from "./base.js";

//#region src/errors/initialization.ts
var InitializationError;
var init_initialization = __esmMin((() => {
	init_base();
	InitializationError = class InitializationError extends AppKitError {
		code = "INITIALIZATION_ERROR";
		statusCode = 500;
		isRetryable = true;
		/**
		* Create an initialization error for a service that is not ready
		*/
		static notInitialized(serviceName, hint) {
			return new InitializationError(hint ? `${serviceName} not initialized. ${hint}` : `${serviceName} not initialized`, { context: { service: serviceName } });
		}
		/**
		* Create an initialization error for setup failure
		*/
		static setupFailed(component, cause) {
			return new InitializationError(`Failed to initialize ${component}`, {
				cause,
				context: { component }
			});
		}
		/**
		* Create an initialization error for migration failure
		*/
		static migrationFailed(cause) {
			return new InitializationError("Error in running migrations for persistent storage", { cause });
		}
	};
}));

//#endregion
init_initialization();
export { InitializationError, init_initialization };
