// @databricks/appkit/dist/errors/server.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { AppKitError, init_base } from "./base.js";

//#region src/errors/server.ts
var ServerError;
var init_server = __esmMin((() => {
	init_base();
	ServerError = class ServerError extends AppKitError {
		code = "SERVER_ERROR";
		statusCode = 500;
		isRetryable = false;
		/**
		* Create a server error for autoStart conflict
		*/
		static autoStartConflict(operation) {
			return new ServerError(`Cannot ${operation} when autoStart is true`, { context: { operation } });
		}
		/**
		* Create a server error for server not started
		*/
		static notStarted() {
			return new ServerError("Server not started. Please start the server first by calling the start() method");
		}
		/**
		* Create a server error for Vite dev server not initialized
		*/
		static viteNotInitialized() {
			return new ServerError("Vite dev server not initialized");
		}
		/**
		* Create a server error for missing client directory
		*/
		static clientDirectoryNotFound(searchedPaths) {
			return new ServerError(`Could not find client directory. Searched for vite.config.ts/js + index.html in: ${searchedPaths.join(", ")}`, { context: { searchedPaths } });
		}
	};
}));

//#endregion
init_server();
export { ServerError, init_server };
