// @databricks/appkit/dist/errors/tunnel.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { AppKitError, init_base } from "./base.js";

//#region src/errors/tunnel.ts
var TunnelError;
var init_tunnel = __esmMin((() => {
	init_base();
	TunnelError = class TunnelError extends AppKitError {
		code = "TUNNEL_ERROR";
		statusCode = 502;
		isRetryable = true;
		/**
		* Create a tunnel error for missing tunnel getter
		*/
		static getterNotRegistered() {
			return new TunnelError("Tunnel getter not registered for DevFileReader singleton");
		}
		/**
		* Create a tunnel error for no available connection
		*/
		static noConnection() {
			return new TunnelError("No tunnel connection available for file read");
		}
		/**
		* Create a tunnel error for asset fetch failure
		*/
		static fetchFailed(path, cause) {
			return new TunnelError("Failed to fetch asset", {
				cause,
				context: { path }
			});
		}
		/**
		* Create a tunnel error for message parsing failure
		*/
		static parseError(messageType, cause) {
			return new TunnelError(`Failed to parse ${messageType} message`, {
				cause,
				context: { messageType }
			});
		}
	};
}));

//#endregion
init_tunnel();
export { TunnelError, init_tunnel };
