// @databricks/appkit/dist/errors/validation.js
import { __esmMin } from "../_virtual/_rolldown/runtime.js";
import { AppKitError, init_base } from "./base.js";

//#region src/errors/validation.ts
var ValidationError;
var init_validation = __esmMin((() => {
	init_base();
	ValidationError = class ValidationError extends AppKitError {
		code = "VALIDATION_ERROR";
		statusCode = 400;
		isRetryable = false;
		/**
		* Create a validation error for a missing required field
		*/
		static missingField(fieldName) {
			return new ValidationError(`Missing required field: ${fieldName}`, { context: { field: fieldName } });
		}
		/**
		* Create a validation error for an invalid field value.
		* Note: The actual value is not stored in context for security reasons.
		* Only the value's type is recorded.
		*/
		static invalidValue(fieldName, value, expected) {
			return new ValidationError(expected ? `Invalid value for ${fieldName}: expected ${expected}` : `Invalid value for ${fieldName}`, { context: {
				field: fieldName,
				valueType: value === null ? "null" : typeof value,
				expected
			} });
		}
		/**
		* Create a validation error for missing environment variables
		*/
		static missingEnvVars(vars) {
			return new ValidationError(`Missing required environment variables: ${vars.join(", ")}`, { context: { missingVars: vars } });
		}
	};
}));

//#endregion
init_validation();
export { ValidationError, init_validation };
